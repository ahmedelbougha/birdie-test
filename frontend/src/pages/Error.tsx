import { StyledError } from '../components';
import { Error } from '../types';

export default function ErrorPage({
  error = {},
  errorStatus,
}: Error): JSX.Element {
  let statusText = '';
  if (errorStatus) {
    // statusText = error.statusText;
  }
  return (
    <StyledError>
      <div>
        <h1>Oops!</h1>
        <p>Sorry, an unexpected error has occurred.</p>
        <p>
          <i>{statusText}</i>
        </p>
      </div>
    </StyledError>
  );
}
