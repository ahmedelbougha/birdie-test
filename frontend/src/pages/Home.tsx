import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { Button, Card, CardsWrapper, StyledError } from "../components";
import { RootState } from '../store';
import { getListRecipients } from '../store/actions';
import { CareRecipient } from '../types';

/**
 * Home page component
 * @returns JSX.Element
 */
function Home(): JSX.Element | null {
  const dispatch = useDispatch();

  const recipients  = useSelector((state: RootState): CareRecipient[] => {
    return state.recipients.list;
  });

  useEffect(() => {
    // get the list of recipients
    if (!recipients) {
      dispatch(getListRecipients());
    }
  }, [dispatch, recipients]);


  if (recipients === null) {
    return null;
  }

  // in case of successful call but no care recipient
  if (recipients.length === 0) {
    return (
      <StyledError>
        <h2>There's no care recipients records in the time being!</h2>
      </StyledError>
    );
  }

  return (
    <>
      <h2>Welcome to Birdie customer portal.</h2>
      <CardsWrapper>
        {recipients.map((recipient: CareRecipient) => (
          <Card
            title={`Mr./Mrs. ${recipient.recipientId.substring(0, 5)}`}
            image=""
            key={recipient.recipientId}
          >
            <p>Birdie Client</p>
            <Link to={`/dashboard/${recipient.recipientId}`}>
              <Button> Details </Button>
            </Link>
          </Card>
        ))}
      </CardsWrapper>
    </>
  );
}

export default Home;
