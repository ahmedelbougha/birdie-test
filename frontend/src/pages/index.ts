export {default as ErrorPage} from "./Error";
export {default as DashboardPage} from "./Dashboard";
export {default as HomePage} from "./Home";

