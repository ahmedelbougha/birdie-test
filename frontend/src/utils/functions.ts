import axios, { AxiosResponse } from "axios";
import { EventTypes } from './constants';


/**
 * General API Requester
 * @returns Promise<AxiosResponse<any, any>>
 */
 export function apiRequest(path: string): Promise<AxiosResponse<any, any>> {
  if (path.length === 0) {
    throw new Error("Request path cannot be empty!");
  }
  return axios.request({
    method: "GET",
    url: `${process.env.REACT_APP_BACKEND_URL}/${path}`,
  });
}


export function buildEventTableEvents() {
  let eventTypes = [];

  // get the main keys of event types in order to use them to display events counts
  // in the EventTable
  const eventTypeKeys = Object.keys(EventTypes);

  for (let i = 0; i < eventTypeKeys.length; i += 2) {
    // build eventTypes array with 2 elements at a time
    // to build the EventTable as 2 events with their count per row
    eventTypes.push([eventTypeKeys[i], eventTypeKeys[i + 1]]);
  }

  return eventTypes;
}
