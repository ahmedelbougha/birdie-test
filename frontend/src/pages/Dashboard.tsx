import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router-dom";
import {
  CardsWrapper,
  Container,
  DashboardCard,
  EventTable,
  EventTimeline,
  StyledError
} from "../components";
import { getEventsRecipient, getSummaryRecipient } from "../store/actions";
import { RootState } from "../store/index";
import { Event } from "../types";
import { buildEventTableEvents } from "../utils/functions";

/**
 * Dashboard component to display recipient complete information
 * @returns JSX.Element
 */
function Dashboard(): JSX.Element | null {
  const dispatch = useDispatch();
  // variable to hold the different event types
  let eventTypes: string[][] = buildEventTableEvents();
  let { recipientId } = useParams();

  const { summaryRecipient, eventsRecipient } = useSelector(
    (
      state: RootState
    ): {
      summaryRecipient: {
        recipientId: string;
        recipientSummary: { [key: string]: string };
      };
      eventsRecipient: Event[];
    } => {
      return state.recipients;
    }
  );

  useEffect(() => {
    // run dispatches in case of recipientId is there and
    // no data coming from selectors or data coming from selector
    // contain different recipientId
    if (
      recipientId &&
      (!summaryRecipient || summaryRecipient.recipientId !== recipientId)
    ) {
      // get the summary of recipient events (counts)
      // used in EventTable and using the recipient id in DashboardCard
      // (due to we don't have access to the name of the care recipient)
      dispatch(getSummaryRecipient(recipientId));

      // get detailed events of the recipient
      // used in the EventTimeline
      dispatch(getEventsRecipient(recipientId));
    }
  }, [dispatch, recipientId, summaryRecipient]);

  if (eventsRecipient === null) {
    return null;
  }

  if (eventsRecipient.length === 0) {
    return (
      <StyledError>
        <h2>There's no matching care recipients records!</h2>
      </StyledError>
    );
  }

  return (
    <>
      <h2>Client Info</h2>
      <Container>
        <CardsWrapper>
          <DashboardCard
            title={`Mr./Mrs. ${summaryRecipient.recipientId.substring(0, 5)}`}
            image=""
          />
          <EventTable
            eventTypes={eventTypes}
            summaryTypes={summaryRecipient.recipientSummary}
          />
        </CardsWrapper>
        <EventTimeline events={eventsRecipient} />
      </Container>
    </>
  );
}

export default Dashboard;
