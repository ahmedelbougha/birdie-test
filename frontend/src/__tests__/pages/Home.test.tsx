import { Provider } from "react-redux";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { ThemeProvider } from "styled-components";
import { HomePage } from "../../pages";
import { theme } from "../../utils/constants";

const router = createBrowserRouter([
  {
    // home page
    path: "/",
    element: <HomePage />,
  },
]);

const mockStore = configureStore([]);

describe("Home Page", () => {
  let testStore: any;

  beforeEach(() => {
    testStore = mockStore({
      recipients: {
        list: null,
        summaryRecipient: null,
        eventsRecipient: null,
      },
    });
  });

  it("Home page in initial state", () => {

    const component = renderer.create(
      <Provider store={testStore}>
        <RouterProvider router={router} />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Home page after loading recipients", () => {
    testStore = mockStore({
      recipients: {
        data: [
          {
            recipientId: "df50cac5-293c-490d-a06c-ee26796f850d",
          },
        ],
      },
    });
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <Provider store={testStore}>
          <RouterProvider router={router} />
        </Provider>
      </ThemeProvider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
