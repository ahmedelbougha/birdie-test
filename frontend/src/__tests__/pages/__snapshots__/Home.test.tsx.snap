// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Home page after loading recipients 1`] = `
Array [
  <h2>
    Welcome to Birdie customer portal.
  </h2>,
  <div
    className="sc-hLBbgP jcEtdA"
  >
    <div
      className="sc-gswNZR eVCMhJ"
    >
      <div>
        <img
          alt="Mr./Mrs. df50c"
          src="placeholder.png"
        />
      </div>
      <div>
        <h2>
          Mr./Mrs. df50c
        </h2>
        <div
          className="sc-bcXHqe AOYhz"
        />
        <p>
          Birdie Client
        </p>
        <a
          href="/dashboard/df50cac5-293c-490d-a06c-ee26796f850d"
          onClick={[Function]}
        >
          <button
            className="sc-kDvujY fuLqXM"
          >
             Details 
          </button>
        </a>
      </div>
    </div>
  </div>,
]
`;

exports[`Home Page Home page in initial state 1`] = `null`;
